### PR TITLE
New Backend Merge

### PR DESCRIPTION
**Pull request by kabdelhak**
Merges the new backend into master

 - experimental flag --newBackend activates it (implies newInst, -nfScalarize and arrayConnect)

### CODE REVIEW

File: OMCompiler/Compiler/Util/UnorderedMap.mo [hunk 1]
```diff
@@ -695,15 +721,16 @@ protected
...
-    hash := hashfn(key, Vector.size(map.buckets));
-    bucket := Vector.get(map.buckets, hash + 1);
-
-    for i in bucket loop
-      if eqfn(key, Vector.getNoBounds(map.keys, i)) then
-        index := i;
-        break;
-      end if;
-    end for;
+    if Vector.size(map.buckets) > 0 then
```
Comment (perost) [inline OMCompiler/Compiler/Util/UnorderedMap.mo]: `hash` never gets a value here if there are no buckets. It might be better to just not allow creating such hashtables.

File: OMCompiler/Compiler/NFFrontEnd/NFType.mo [hunk 2]
```diff
@@ -1303,6 +1319,19 @@ public
     str := stringAppendList(strl);
   end subscriptedTypeName;
 
+  function addDimensions
```
Comment (perost) [inline OMCompiler/Compiler/NFFrontEnd/NFType.mo]: There's already `liftArrayLeftList` that does this.

File: OMCompiler/Compiler/Util/UnorderedMap.mo (hunk 1 above)
Comment (kabdelhak) [reply OMCompiler/Compiler/Util/UnorderedMap.mo]: thats true, for now i just pulled the hash computation out of the function

File: OMCompiler/Compiler/NFFrontEnd/NFType.mo (hunk 2 above)
Comment (kabdelhak) [reply OMCompiler/Compiler/NFFrontEnd/NFType.mo]: okay i replaced it!

File: OMCompiler/Compiler/NFFrontEnd/NFSimplifyExp.mo [hunk 3]
```diff
@@ -84,22 +99,22 @@ algorithm
...
-    case Expression.UNARY() then simplifyUnary(exp);
-    case Expression.LBINARY() then simplifyLogicBinary(exp);
-    case Expression.LUNARY() then simplifyLogicUnary(exp);
-    case Expression.RELATION() then simplifyRelation(exp);
-    case Expression.IF() then simplifyIf(exp);
-    case Expression.CAST() then simplifyCast(simplify(exp.exp), exp.ty);
-    case Expression.UNBOX() then Expression.UNBOX(simplify(exp.exp), exp.ty);
-    case Expression.SUBSCRIPTED_EXP() then simplifySubscriptedExp(exp);
-    case Expression.TUPLE_ELEMENT() then simplifyTupleElement(exp);
-    case Expression.RECORD_ELEMENT() then simplifyRecordElement(exp);
```
[the remark this answers is not in the document]
Comment (kabdelhak) [reply OMCompiler/Compiler/NFFrontEnd/NFSimplifyExp.mo]: oops :) i'll put it back!

File: OMCompiler/Compiler/Util/UnorderedMap.mo (hunk 1 above)
Comment (perost) [reply OMCompiler/Compiler/Util/UnorderedMap.mo:725]: I thought it was the modulo with 0 that the hash computation will do that was the issue, but maybe that's fine in MetaModelica.

**Comment by phannebohm** [comment]
The last one `simulation/modelica/start_value_selection/ticket5807.mos` looks like it's trailing whitespace :roll_eyes:

**Comment by kabdelhak** [comment]
> The last one `simulation/modelica/start_value_selection/ticket5807.mos` looks like it's trailing whitespace roll_eyes

yea, but it seems to be a git issue. If i fix the testsuite and commit it, git reverts the changes. I tried commiting with `-n` maybe that suffices

**Comment by perost** [comment]
> The last one `simulation/modelica/start_value_selection/ticket5807.mos` looks like it's trailing whitespace roll_eyes

The issue seems to be that the debug functions in [nonlinearSolverHomotopy.c](https://github.com/OpenModelica/OpenModelica/blob/master/OMCompiler/SimulationRuntime/c/simulation/solver/nonlinearSolverHomotopy.c) prints a space after every number including the last one, which git then removes. Maybe we should change `rtest` to not care about trailing whitespace, it's a confusing issue whenever it pops up.

**Comment by phannebohm** [comment]
I just opened #8525, we shouldn't print trailing whitespace in the first place.

**Comment by perost** [comment]
> I just opened #8525, we shouldn't print trailing whitespace in the first place.

Yeah, I'm a bit torn. On one hand we don't want trailing whitespace, on the other hand it can be very confusing when we do have it and git removes it. Maybe we should have rtest just reject tests with trailing whitespace to avoid the situation, or at least give a warning.

**Comment by perost** [comment]
I opened #8526 that makes `rtest` warn about trailing whitespace when baselining a test, so baselining e.g. ticket5807.mos now gives:
```Modelica
Warning: trailing whitespace:
  |                 | |       | |               -2 
                                                  ^
Warning: trailing whitespace:
  |                 | |       | |                1 
                                                  ^
Warning: trailing whitespace:
  |                 | |       | |                2                1 
                                                                   ^
Warning: trailing whitespace:
  |                 | |       | |               -2 
                                                  ^
Warning: trailing whitespace:
  |                 | |       | | 0 
                                   ^
Warning: trailing whitespace:
  |                 | |       | | 0 1 
                                     ^
Warning: trailing whitespace:
  |                 | |       | |               -0                1 
                                                                   ^
Warning: trailing whitespace:
  |                 | |       | |                0 
                                                  ^
Warning: trailing whitespace:
  |                 | |       | |                0 
                                                  ^
Warning: trailing whitespace:
  |                 | |       | |               -2 
                                                  ^
```
Hopefully that will at least deter people from adding new tests with trailing whitespace.

**Review by mahge — APPROVED** [review]
